### ISSUE TITLE
Issues with "Fetching and Parsing Data from the Web with OpenRefine"

### ISSUE THREAD

**Issue by fredericofreitas-ncsu** [programminghistorian/jekyll#2687]
Lesson: [Fetching and Parsing Data from the Web with OpenRefine](https://programminghistorian.org/en/lessons/fetch-and-parse-data-with-openrefine)

System: MacOS 12.5.1
Software: OpenRevine 3.6.1

I used Examples 1 and 2 in class, and I found a few small bugs. I was able to identify the issues and correct them.

**Example 1: Fetching and Parsing HTML**

In this tutorial, some of the functions in the **Extract Information with Array Functions** subsection are not working. Where they have `<br />` in their function, it should be `<br>`. Thus:

```
New column "number"
Instead of: value.split("<br />”)[0].trim()
Use: value.split("<br>”)[0].trim()

New column "first"
Instead of: value.split("<br />")[1].trim()
Use: value.split("<br>")[1].trim()

New column "text"
Instead of: forEach(value.split("<br />"), line, line.trim()).slice(1).join("\n")
Use: forEach(value.split("<br>"), line, line.trim()).slice(1).join("\n")

New column 'last"
Instead of: forEach(value.split("<br />"), line, line.trim()).slice(-3).join("\n")
Use: forEach(value.split("<br>"), line, line.trim()).slice(-3).join("\n")
```

**Example 2: Fetching and Parsing HTML**

The text that is scraped from the front pages of the newspaper now has some extra | characters, which makes things not work. To solve this, we have to change the character used for **join**:

```
Parse JSON to Get Items
Instead of: value.parseJson()['items'].join("|||")
Use: value.parseJson()['items'].join("|^^^|")

Split Multivalued Cells
Instead of using as separator the character |
Use as separators the characters ^^^|
```

**Comment by acrymble** [comment]
Thanks for reporting this @fredericofreitas-ncsu ! We'll take a look

**Comment by anisa-hawes** [comment]
Hello @fredericofreitas-ncsu. Thank you for getting in touch. I've tested the lesson Examples 1 and 2. I've adjusted the original code as you suggest, replacing `<br />` with `<br>` and can confirm that this works well. I'm happy to update the lesson with this change. Here are my final columns for Example 1: Sonnets 

![Screenshot 2022-10-12 at 15 38 52](https://user-images.githubusercontent.com/87070441/195381317-87535e91-65e7-4272-b218-0185a836219c.png)

But I'm less certain about your suggested edit to the expressions used within Example 3: ChronAm. The line breaks within by final _text_ column are not as I expected (comparing to the lesson's screenshots). 

![Screenshot 2022-10-12 at 16 10 03](https://user-images.githubusercontent.com/87070441/195383901-23c1d204-82c5-4b65-9031-a0eb274431c0.png)

I'll work through that second example again and report back.

**Comment by anisa-hawes** [comment]
I've re-tested Example 3, adjusting the character used for `join`:

At step [Parse JSON to Get Items](https://programminghistorian.org/en/lessons/fetch-and-parse-data-with-openrefine#parse-json-to-get-items) I used the expression: `value.parseJson()['items'].join("^^^")`. N.B. Without any pipes, `join("^^^")` instead of `join("|^^^|")`.

![Screenshot 2022-10-12 at 16 49 16](https://user-images.githubusercontent.com/87070441/195392494-600d4ee4-8b56-40ce-afb7-f5116f9fb222.png)

At the following step [Split Multivalued Cells](https://programminghistorian.org/en/lessons/fetch-and-parse-data-with-openrefine#split-multivalued-cells) I entered the separator `^^^`. N.B. Again, without any pipes, `^^^` instead of `^^^|` ([as given above](https://github.com/programminghistorian/jekyll/issues/2687#issue-1385916480)).

![Screenshot 2022-10-12 at 16 50 32](https://user-images.githubusercontent.com/87070441/195393123-36fe7480-8c17-4a81-90d3-41783ba6f5eb.png)

My final columns appear to be the same as in my previous test, with strange line-breaks. 

![Screenshot 2022-10-12 at 16 59 21](https://user-images.githubusercontent.com/87070441/195395759-2b3d191d-4906-457a-83d9-c5bd5891219e.png)

Everything else seems to be as it should. I'm able to transform the dates to strings, and to extend the data with URL queries as shown. 

I'll prepare the updates, but wonder if I am still missing something/doing something odd here that's causing the unexpected line breaks?

**Comment by fredericofreitas-ncsu** [comment]
I am not sure what is causing the line breaks… As you know, the problem I had identified was in extracting the text of the individual different newspaper editions. They changed something at the Library of Congress and now the OCR’ed text was coming with pipe characters, which messed things up in the original parsing process (which used pipe characters as separators). So my idea was to use a unique combination of characters that was unlike to appear in the OCR, i.e., ^^^. It could be any other unusual combination of characters. Now, the line breaks… I am not sure what is going on. 


Frederico Freitas, Ph.D.
Associate Professor of History
North Carolina State University
919-515-3307
fredericofreitas.org ***@***.***

Author of Nationalizing Nature: Iguazu Falls and National Parks at the Brazil-Argentina Border <http://www.cambridge.org/9781108844833> (Cambridge University Press, 2021).

**Comment by anisa-hawes** [comment]
Thank you, @fredericofreitas-ncsu. That makes sense, and your suggestion certainly works. Using your method, I can confirm that my project table has 20 rows following the [Split Multivalued Cells](https://programminghistorian.org/en/lessons/fetch-and-parse-data-with-openrefine#split-multivalued-cells) action, which is right. 

Looking more closely, I can see that the OCR output now introduces random new lines `\n` throughout the text, which aren't visible in the lesson's original screenshots. I don't know whether it's possible to filter them out using GREL? 

This is unsophisticated, but I tried `forEach(value.split("\n"), line, line.trim()).slice(1).join(" ")` and that made it look a bit better:

![Screenshot 2022-10-12 at 19 53 24](https://user-images.githubusercontent.com/87070441/195424971-2eea01fa-091e-443f-bbcc-ff4fe0029ab6.png)

**Comment by fredericofreitas-ncsu** [comment]
I am not sure if it is the best way of doing this, but maybe using replace and replacing “\n” with “” [nothing]?

.replace(“\n", "")

Not sure if that would work, though. 


Frederico Freitas, Ph.D.
Associate Professor of History
North Carolina State University
919-515-3307
fredericofreitas.org ***@***.***

Author of Nationalizing Nature: Iguazu Falls and National Parks at the Brazil-Argentina Border <http://www.cambridge.org/9781108844833> (Cambridge University Press, 2021).

**Comment by anisa-hawes** [comment]
At l.129 of the .md file, I notice that the figure caption reads: _Each sonnet is a `<p>` with lines separated by `<br />`_ and indeed, that is the case when I View Page Source so I am not adjusting this figure image or caption.

![Screenshot 2022-11-03 at 18 07 53](https://user-images.githubusercontent.com/87070441/199970530-e752ba9f-1d04-447c-9571-33e84f3a8edb.png)

However, in my tests I found @fredericofreitas-ncsu is correct to advise that `<br />` needs to be replaced with `<br>` in Expression boxes throughout.

--

I've added an alert-info box at l.464 to advise users that errant line breaks can be removed using the Expression `value.replace("\n","")`. Thank you @fredericofreitas-ncsu, this worked perfectly in my tests: 

Before:
![Screenshot 2022-11-03 at 18 49 44](https://user-images.githubusercontent.com/87070441/199971062-ca2417dd-d5b0-4b31-94f6-085bf3953226.png)

Testing the Expression:
![Screenshot 2022-11-03 at 18 52 34](https://user-images.githubusercontent.com/87070441/199971305-cd698003-e6be-4405-955c-e11604383e54.png)

After:
![Screenshot 2022-11-03 at 18 54 10](https://user-images.githubusercontent.com/87070441/199971183-8138fab2-6364-4112-89db-93f69ee51bfe.png)

--
Edits are being prepared in PR #2732 🙂
